### PR TITLE
[master] ZIL-5011: move more dependencies into Dockerfile.deps

### PR DESCRIPTION
**Pull request by yaron-zilliqa**
## Description

* Removing the REGISTRY argument as it's not needed (scilla & the dependencies plan push to DockerHub).
* Move more things into the dependencies docker image so that Zilliqa builds take a bit less time.
* Use the Ninja system which seems to be slightly more efficient.
* Simply more scripts to use build.sh instead of repeating the same commands in multiple places.
* Create more Dockerfile so that we can benefit better from multi-stage builds.
* Build a ccache Docker image that can be reused to drastically speed up compilation.

## Backward Compatibility

- [x] This is not a breaking change
- [ ] This is a breaking change

### Implementation

- [x] **ready for review**

### Integration Test (Core Team)

- [ ] local machine test

- [ ] small-scale cloud test

### CODE REVIEW

**Comment by JamesHinshelwood** [comment]
> Removing the REGISTRY argument as it's not needed (scilla & the dependencies plan push to DockerHub).

@mauromedda added this not long ago, so check we definitely don't. In particular, I could imagine we might have pre-release builds of deps and scilla that aren't on dockerhub yet, but are in our private registry?

**Comment by yaron-zilliqa** [comment]
> > Removing the REGISTRY argument as it's not needed (scilla & the dependencies plan push to DockerHub).
> 
> @mauromedda added this not long ago, so check we definitely don't. In particular, I could imagine we might have pre-release builds of deps and scilla that aren't on dockerhub yet, but are in our private registry?

I know he did and this was done because some Jenkins jobs pushed to ECR and some to DockerHub. The dependency job now pushes and DockerHub and yet in some cases when the registry was pushed set to ECR it was apparently using a different deps build.
A bit confusing and inconsistent and pretty error prone, but if needed we can keep it.  Other than this, it's not really used at the moment.

**Comment by JamesHinshelwood** [comment]
> > > Removing the REGISTRY argument as it's not needed (scilla & the dependencies plan push to DockerHub).
> > 
> > 
> > @mauromedda added this not long ago, so check we definitely don't. In particular, I could imagine we might have pre-release builds of deps and scilla that aren't on dockerhub yet, but are in our private registry?
> 
> I know he did and this was done because some Jenkins jobs pushed to ECR and some to DockerHub. The dependency job now pushes and DockerHub and yet in some cases when the registry was set to ECR it was apparently using a different (old) deps build. A bit confusing and inconsistent and pretty error prone, but if needed we can keep it. Other than this, it's not really used at the moment.

Yeah fair points, makes sense. (though my other comment about removing `.deps` entirely still stands :stuck_out_tongue: )

File: docker/Dockerfile.deps [hunk 1]
```diff
@@ -15,7 +15,8 @@
 #
 
 # start from a new ubuntu environment as builder for zilliqa, make sure the deps is consistent with those in the runner image
-FROM ubuntu:20.04 AS builder
```
[the remark this answers is not in the document]
Comment (yaron-zilliqa) [reply docker/Dockerfile.deps:18]: Also for scilla, but of course we can't combine. The whole point is *not* to build the dependencies each time but let other images reuse it.

**Comment by yaron-zilliqa** [comment]
> Yeah fair points, makes sense. (though my other comment about removing `.deps` entirely still stands stuck_out_tongue )

I replied that we can't just do it because the whole point is not to rebuild the dependencies each time. Perhaps you meant something else.

File: docker/Dockerfile.deps (hunk 1 above)
[the remark this answers is not in the document]
Comment (JamesHinshelwood) [reply docker/Dockerfile.deps:18]: We could use a multi-stage build and rely on docker caching.

[the remark this answers is not in the document]
Comment (yaron-zilliqa) [reply docker/Dockerfile.deps:18]: We already use multi-stage builds and caching (see Dockerfile). In some situations unfortunately, we use arguments which is equivalent to an if-else and prevents Docker from caching the layer. Anyway, not sure what your idea is.

**Comment by yaron-zilliqa** [comment]
> I think this looks okay, but I don't think I'm entirely convinced by the plan of having lots of different base images and combining them together. IMO this makes the dependencies of each image hard to keep track of.
> 
> As far as I understand, the goal here is to reduce docker build times by caching stuff in base images? I guess I would argue that we should try to do this with Docker's native caching of layers, rather than attempting to do it ourselves on an image-level. That way we won't run into dirty cache issues, where we fail to update base images.
> 
> I am not super confident either way though, so I'd be curious to know what others think? @rrw-zilliqa @mauromedda ?

Not a huge fan of this myself either, but at the moment I'm not sure there's an easier way of doing this. Not sure if we can rely entirely on layer caching. By the way, some of the layers have been changed in order to create smaller images. My understanding is that we're also looking into alternatives to our current Docker so I view this as an intermediary solution. If we want to borrow in the future - we can.

An important part is actually the ccache image which is meant to speed up compilation (currently it shaves several minutes off our builds). This might be redundant in the future if we can provide a common storage that can be used easily. (it's possible to do with ccache but trickier and I'd rather start with this). This image might be completely redundant in the future.

**Review by mauromedda — APPROVED** [review]
LGTM

**Pull request merged**